lib/bolt/cli.rb

```diff
@@ -564,19 +564,121 @@ def generate_types
+      # might not have installed any modules yet.
+      if config.exist?
+        @logger.warn "Found existing project directory at #{boltdir}"
+      end
 
-      ok ? 0 : 1
+      # Create the project directory
+      FileUtils.mkdir_p(boltdir)
+
+      # Bless the project directory as a boltdir
+      if FileUtils.touch(config)
+        outputter.print_message "Successfully created Bolt project at #{boltdir}"
+      else
+        raise Bolt::FileError.new("Could not create Bolt project directory at #{boltdir}", nil)
+      end
+
+      # Write the generated Puppetfile to the fancy new boltdir
+      if puppetfile_specs
+        if File.write(puppetfile, puppetfile_specs.join("\n"))
+          outputter.print_message "Successfully created Puppetfile at #{puppetfile}"
+          # Install the modules from our shiny new Puppetfile
+          if install_puppetfile({}, puppetfile, modulepath)
+            outputter.print_message "Successfully installed module#{plural} #{options[:modules].join(', ')}"
+          else
+            raise Bolt::CLIError, "Could not install module#{plural} #{options[:modules].join(', ')}"
+          end
+        else
+          raise Bolt::FileError.new("Could not write Puppetfile at #{puppetfile}", nil)
+        end
+      end
+
+      0
+    end
+
+    # Resolves Puppetfile specs from user-specified modules and dependencies resolved
+    # by the puppetfile-resolver gem.
+    def resolve_puppetfile_specs
+      require 'puppetfile-resolver'
+
+      # Build the document model from the module names, defaulting to the latest version of each module
+      model = PuppetfileResolver::Puppetfile::Document.new('')
+      options[:modules].each do |mod_name|
+        model.add_module(
+          PuppetfileResolver::Puppetfile::ForgeModule.new(mod_name).tap { |mod| mod.version = :latest }
+        )
+      end
+
+      # Make sure the Puppetfile model is valid
+      unless model.valid?
+        plural = options[:modules].count == 1 ? '' : 's'
+        raise Bolt::ValidationError,
+              "Unable to resolve dependencies for module#{plural} #{options[:modules].join(', ')}"
+      end
+
+      # Create the resolver using the Puppetfile model. nil disables Puppet version restrictions.
+      resolver = PuppetfileResolver::Resolver.new(model, nil)
+
+      # Configure and resolve the dependency graph
+      result = resolver.resolve(
+        cache:                 nil,
+        ui:                    nil,
+        module_paths:          [],
+        allow_missing_modules: true
+      )
+
+      # Validate that the modules exist
+      missing_graph = result.specifications.select do |_name, spec|
+        spec.instance_of? PuppetfileResolver::Models::MissingModuleSpecification
+      end
+
+      if missing_graph.any?
+        titles = model.modules.each_with_object({}) do |mod, acc|
+          acc[mod.name] = mod.title
+        end
+
+        names  = titles.slice(*missing_graph.keys).values
```

Comment:
I believe this is equivalent to `names = titles.values_at(*missing_graph.keys)`.